cw/capture.py

```diff
@@ -1169,168 +1171,315 @@ def capture_otbn_vertical(ot, ktp, fw_bin, pll_frequency, capture_cfg, device_cf
+                f'Hint: set plaintext len={seed_bytes} in the configuration file.'
+            )
 
-        ot.scope.adc.offset = capture_cfg["offset"]
+        # select the otbn app on the device (0 -> keygen, 1 -> modinv)
+        ot.target.simpleserial_write("a", bytearray([0x00]))
+        time.sleep(0.3)
 
-        if capture_cfg["masks_off"] is True:
-            # Use a constant mask for each trace
-            mask = bytearray(capture_cfg["plain_text_len_bytes"])  # all zeros
-        else:
-            # Generate a new random mask for each trace.
-            mask = ktp.next_text()
-
-        tqdm.write("Starting new trace....")
-        tqdm.write(f'mask   = {mask.hex()}')
+        # Seed the RNG and generate a random fixed seed for all traces of the keygen operation
+        random.seed(capture_cfg["batch_prng_seed"])
 
+        # Generate fixed constants for all traces of the keygen operation.
         if capture_cfg["test_type"] == 'KEY':
-            # In fixed-vs-random KEY mode, the fixed set of measurements is
-            # generated using the fixed 320 bit seed. The random set of
-            # measurements is generated in two steps:
-            #    1. Choose a random 256 bit number r
-            #    2. Compute the seed as (C + r) where C is the fixed 320 bit
-            #       constant. Note that in this case the used key is equal to
-            #       (C + r) mod curve_order_n
-            if sample_fixed:
-                seed_used = seed_fixed
-                expected_key = expected_fixed_key
-            else:
-                ktp.key_len = key_bytes
-                random_number = ktp.next_key()
-                ktp.key_len = seed_bytes
-                seed_used_int = int.from_bytes(C, byteorder='little') + \
-                    int.from_bytes(random_number, byteorder='little')
-                seed_used = seed_used_int.to_bytes(seed_bytes, byteorder='little')
-                expected_key = int.from_bytes(seed_used, byteorder='little') % curve_order_n
+            # In fixed-vs-random KEY mode we use two fixed constants:
+            #    1. C - a 320 bit constant redundancy
+            #    2. fixed_number - a 256 bit number used to derive the fixed key
+            #                      for the fixed set of measurements. Note that in
+            #                      this set, the fixed key is equal to
+            #                      (C + fixed_number) mod curve_order_n
+            C = ktp.next_key()
+            if len(C) != seed_bytes:
+                raise ValueError(f'Fixed seed length is {len(C)}, expected {seed_bytes}')
+            ktp.key_len = key_bytes
+            fixed_number = ktp.next_key()
+            if len(fixed_number) != key_bytes:
+                raise ValueError(f'Fixed key length is {len(fixed_number)}, expected {key_bytes}')
+            ktp.key_len = seed_bytes
+
+            seed_fixed_int = int.from_bytes(C, byteorder='little') + \
```

Comment:
This might become a 321-bit number, right? Should we substract n from it if it becomes a 321-bit number?

Reply:
Theoretically yes I would say. Might be a good idea to subtract n.
But maybe it would be a good idea to do this in a separate PR?

Reply:
Yes, this seems to be a pre-existing issue.
However, the only way an overflow could happen is if C > 2^320 - 2^64, which is probably not the constant we want to use.
If we decide to add a check for this I would prefer to raise an error rather than subtract `n`. Either way, this PR doesn't need to change it.